File: client/di_core/LakeHouse/Domain/LakeJob/GitCloneInfo/GitCloneInfo.ts
import { GitCloneInfos } from '@core/LakeHouse/Domain/LakeJob/GitCloneInfo/GitCloneInfos';
import { SSHInfo } from '@core/LakeHouse/Domain/LakeJob/GitCloneInfo/SSHInfo';
import { HTTPSInfo } from '@core/LakeHouse/Domain/LakeJob/GitCloneInfo/HTTPSInfo';
import { UnsupportedException } from '@core/domain/Exception/UnsupportedException';

export abstract class GitCloneInfo {
  abstract className: GitCloneInfos;

  abstract isValid(): boolean;

  static fromObject(obj: any): GitCloneInfo {
    switch (obj.className as GitCloneInfos) {
      case GitCloneInfos.HTTPS:
        return new HTTPSInfo(obj.url, obj.username, obj.password);
      case GitCloneInfos.SSH:
        return new SSHInfo(obj.url, obj.privateKey);
      default:
        throw new UnsupportedException(`Unsupported git clone type ${obj.className}`);
    }
  }
}

File: job-scheduler/src/main/scala/datainsider/lakescheduler/domain/GitCloneInfo.scala
package datainsider.lakescheduler.domain

import com.fasterxml.jackson.annotation.JsonSubTypes.Type
import com.fasterxml.jackson.annotation.{JsonSubTypes, JsonTypeInfo}

@JsonTypeInfo(
  use = JsonTypeInfo.Id.NAME,
  include = JsonTypeInfo.As.PROPERTY,
  property = "class_name"
)
@JsonSubTypes(
  Array(
    new Type(value = classOf[HttpCloneInfo], name = "http_clone_info"),
    new Type(value = classOf[SshCloneInfo], name = "ssh_clone_info")
  )
)
abstract class GitCloneInfo

case class HttpCloneInfo(url: String, username: String, password: String) extends GitCloneInfo

case class SshCloneInfo(url: String, privateKey: String) extends GitCloneInfo
